### ISSUE TITLE
SQL Server: `prisma generate` fails with error "memory allocation of [86GB] failed"

### ISSUE THREAD

**Issue by peteralbert** [prisma/prisma#10621]
### Bug description

We just upgraded from Prisma 2.30.2 to 3.6.0.

Running `prisma generate` fails with this error message (after a very long time, 30mins plus!):

```
$ npx prisma generate
Prisma schema loaded from prisma\schema.prisma
memory allocation of 42949672960 bytes failed
```

The same problem occurs with Prisma 3.0.2

### How to reproduce

I created a small repo with our schema that reproduces the issue: https://github.com/peteralbert/prisma-generate-issue

### Expected behavior

Command should generate Prisma client without request 86GB of RAM

### Prisma information

See [schema in the repo](https://github.com/peteralbert/prisma-generate-issue/blob/master/prisma/schema.prisma)

### Environment & setup

- OS: Windows Server 2019 Standard
- Database: SQL Server
- Node.js version: v14.15.1

### Prisma Version

```
prisma                  : 3.6.0
@prisma/client          : 3.6.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at ..\..\node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli dc520b92b1ebb2d28dc3161f9f82e875bd35d727
(at ..\..\node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at ..\..\node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt dc520b92b1ebb2d28dc3161f9f82e875bd35d727 (at ..\..\node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : dc520b92b1ebb2d28dc3161f9f82e875bd35d727
Studio                  : 0.440.0
```

**Comment by peteralbert** [comment]
Further investigating the issue - it seems that it is caused by the "cycle detection" - the schema does not specify any `onDelete` or `onUpdate` actions yet. In a smaller version, `prisma generate` shows error messages of detected cycles.

I suspect that there are too many cycles and hence the "cycle detection" requires too much memory...

Will add the actions to all relationships and see...

**Comment by pantharshit00** [comment]
Thanks for reporting. I can reproduce this. 

<img width="920" alt="Screenshot 2021-12-08 at 12 07 07 PM" src="https://user-images.githubusercontent.com/22195362/145198364-e55f297e-94f1-4df9-97b0-1349606ab13f.png">

**Comment by pimeys** [comment]
Hey @peteralbert. It's not in cycle detection, but in the cascading paths validation we have an endless cycle. I'm currently making a minimal reproduction, but meanwhile could you `db pull` with the latest prisma version and see if it modifies the data model to a state that doesn't cycle anymore? Your old data model didn't have actions written down, and I highly doubt you have cycles or multiple cascading paths between relations in your SQL Server database, so the generated new data model with `db pull` should not trigger a loop.

This needs a fast fix, so I'm looking into doing it today.

**Comment by peteralbert** [comment]
@pimeys `db pull` shows similar behavior - it is now running since >5mins and grabbing more and more memory: 
![image](https://user-images.githubusercontent.com/2182870/145208661-32f73ace-7df4-44ef-9ce1-3a2b66a59c42.png)

I'll see if I can share the DB schema so you can reproduce locally...

**Comment by peteralbert** [comment]
Cancelling `db pull`now, current status was 
![image](https://user-images.githubusercontent.com/2182870/145210674-64fe1736-25af-4e3f-a5df-bc1d095b831e.png)

**Comment by pimeys** [comment]
The SQL schema could be useful for us. You can also send it to schemas@prisma.io if you don't want to include it in the ticket.

**Comment by peteralbert** [comment]
Just shared the deploy script via the above email. Thanks

**Comment by pimeys** [comment]
Thanks. We got it. I'm also close on solving the issue.

**Comment by pimeys** [comment]
I have a fix PR. There were two problems. Both are fixed now.

Happy holidays!

**Comment by pimeys** [comment]
When this is merged and you want to try the new version, remember to `db pull` before doing anything with your current schema to get the referential actions and constraint names to their places.

Here's how it will look like after doing so:

https://gist.github.com/pimeys/9e76b1d149d6eccdd570cc9151d939ce

**Comment by peteralbert** [comment]
Thanks @pimeys ! Looking forward to test/apply it!

**Comment by janpio** [comment]
In a bit this should be available in a `dev` version for testing - will post the exact version number when it happened.

**Comment by peteralbert** [comment]
@janpio  Sorry, did not see the update. Is this part of the released v3.7?

**Comment by janpio** [comment]
Yes indeed, this should now be publicly available in 3.7.0 @peteralbert 👍

**Comment by peteralbert** [comment]
awesome! will verify after xmas break
